x-pack/legacy/plugins/ml/public/application/explorer/actions/load_explorer_data.ts

```diff
@@ -0,0 +1,327 @@
+                )
+              : Promise.resolve(influencers),
+          viewBySwimlaneState: memoizedLoadViewBySwimlane(
+            topFieldValues,
+            {
+              earliest: overallState.overallSwimlaneData.earliest,
+              latest: overallState.overallSwimlaneData.latest,
+            },
+            selectedJobs,
+            viewBySwimlaneFieldName,
+            swimlaneLimit,
+            influencersFilterQuery,
+            noInfluencersConfigured
+          ),
+        }),
+      ({ annotationsData, overallState, tableData }, { influencers, viewBySwimlaneState }) => {
+        return {
+          type: EXPLORER_ACTION.SET_STATE,
+          payload: {
+            annotationsData,
+            influencers,
+            ...overallState,
+            ...viewBySwimlaneState,
+            tableData,
+          },
+        };
+      }
+    ),
+    // do a sanity check against selectedCells. It can happen that a previously
+    // selected lane loaded via URL/AppState is not available anymore.
+    // If filter is active - selectedCell may not be available due to swimlane view by change to filter fieldName
+    // Ok to keep cellSelection in this case
+    map(action => {
+      const { viewBySwimlaneData } = action.payload;
+
+      let clearSelection = false;
+      if (selectedCells !== null && selectedCells.type === SWIMLANE_TYPE.VIEW_BY) {
+        clearSelection =
+          filterActive === false &&
+          !selectedCells.lanes.some((lane: string) => {
+            return viewBySwimlaneData.points.some((point: SwimlanePoint) => {
+              return (
+                point.laneLabel === lane &&
+                point.time >= selectedCells.times[0] &&
+                point.time <= selectedCells.times[1]
+              );
+            });
+          });
+      }
+
+      if (clearSelection === true) {
+        explorerAction$.next({ type: EXPLORER_ACTION.APP_STATE_CLEAR_SELECTION });
+        action.payload = { ...action.payload, ...getClearedSelectedAnomaliesState() };
+      }
+
+      return action;
+    }),
+    // Set the KQL query bar placeholder value
+    map(action => {
+      const { influencers } = action.payload;
+
+      if (influencers !== undefined && !noInfluencersConfigured) {
+        for (const influencerName in influencers) {
+          if (
+            influencers[influencerName][0] &&
+            influencers[influencerName][0].influencerFieldValue
+          ) {
+            action.payload.filterPlaceHolder = i18n.translate(
+              'xpack.ml.explorer.kueryBar.filterPlaceholder',
+              {
+                defaultMessage: 'Filter by influencer fields… ({queryExample})',
+                values: {
+                  queryExample: `${influencerName} : ${influencers[influencerName][0].influencerFieldValue}`,
+                },
+              }
+            );
+            break;
+          }
+        }
+      }
+
+      return action;
+    })
+  );
+}
```

Comment:
overall I find this file too difficult to follow. there is a lot of going in on with creating observable from promises, combining them, some side effects. I'm not sure we need a concept of actions here at all. You just want to fetch data for explorer and map it to a certain format. So it basically might be a combination of observables with some mapping. It'd be easier to test also

Reply:
Thanks for the feedback! I did a cleanup of this file in 6bad98d. I moved a lot of the code that's just transforming data without remote fetching to `explorer_reducer.ts`. The whole page now does way less state update triggers and rerenders. The side-effects are still necessary here because e.g. we want to update the page once the overall swimlane loaded and the go on and load the view-by swimlanes. But the code should be a lot easier to read. It's also no longer based on the action format, I moved that to the place where the observable is consumed in `explorer_dashboard_service.ts`.